### PR TITLE
0.30: Updates to Seat module

### PR DESCRIPTION
**Pull request by PolyMeilex**
closes: #461

- Adds multiseat support
- Brings back WlTouch support

### CODE REVIEW

File: src/wayland/seat/mod.rs [hunk 1]
```diff
@@ -333,77 +361,114 @@ impl<T: 'static> SeatState<T> {
...
+/// This struct gives you access to the control of the
+/// capabilities of the associated seat.
+///
+/// It is directly inserted in the wayland display by its [`new`](Seat::new) method.
+///
+/// This is an handle to the inner logic, it can be cloned.
+///
+/// See module-level documentation for details of use.
+#[derive(Debug)]
+pub struct SeatState<T> {
```
Comment (i509VCB) [inline src/wayland/seat/mod.rs:438]: I think you have the wrong docs here

File: src/wayland/seat/mod.rs [hunk 2]
```diff
@@ -333,77 +361,114 @@ impl<T: 'static> SeatState<T> {
...
+    pd: PhantomData<T>,
+}
+
+impl<T> Default for SeatState<T> {
```
Comment (i509VCB) [inline src/wayland/seat/mod.rs:442]: Probably not needed.

File: src/wayland/seat/mod.rs [hunk 3]
```diff
@@ -333,77 +361,114 @@ impl<T: 'static> SeatState<T> {
...
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T> Clone for SeatState<T> {
+    fn clone(&self) -> Self {
```
Comment (i509VCB) [inline src/wayland/seat/mod.rs:449]: Hmm this seems like the wrong thing to be able to clone.

File: src/wayland/seat/mod.rs (hunk 1 above)
Comment (PolyMeilex) [reply src/wayland/seat/mod.rs:432]: Indeed

File: src/wayland/seat/mod.rs (hunk 2 above)
Comment (PolyMeilex) [reply src/wayland/seat/mod.rs:436]: This was done to make Clippy happy.

File: src/wayland/seat/mod.rs (hunk 3 above)
Comment (PolyMeilex) [reply src/wayland/seat/mod.rs]: Right

File: src/wayland/seat/touch.rs [hunk 4]
```diff
@@ -109,83 +129,103 @@ impl TouchInternal {
...
+{
+    fn request(
+        _state: &mut D,
+        _client: &wayland_server::Client,
+        _resource: &WlTouch,
+        _request: wl_touch::Request,
+        _data: &Self::UserData,
+        _dhandle: &mut DisplayHandle<'_>,
+        _data_init: &mut wayland_server::DataInit<'_, D>,
+    ) {
```
[the remark this answers is not in the document]
Comment (PolyMeilex) [reply src/wayland/seat/touch.rs:218]: Done.

File: src/wayland/seat/mod.rs (hunk 2 above)
Comment (i509VCB) [reply src/wayland/seat/mod.rs:436]: Well we could suppress this on the `new` function via `#[allow(clippy::new_without_default)]`.

Considering other delegate types require the display, `Default` isn't going to do a whole lot for usability on these types.

File: src/wayland/seat/mod.rs [hunk 5]
```diff
@@ -507,40 +567,39 @@ where
...
 
 impl<T: 'static> DelegateGlobalDispatchBase<WlSeat> for SeatState<T> {
-    type GlobalData = ();
+    type GlobalData = SeatGlobalData<T>;
 }
 
 impl<T, D> DelegateGlobalDispatch<WlSeat, D> for SeatState<T>
 where
-    D: GlobalDispatch<WlSeat, GlobalData = ()>,
+    D: GlobalDispatch<WlSeat, GlobalData = <Self as DelegateGlobalDispatchBase<WlSeat>>::GlobalData>,
```
Comment (i509VCB) [inline src/wayland/seat/mod.rs:575]: `Self::GlobalData` doesn't work here?

File: src/wayland/seat/touch.rs [hunk 6]
```diff
@@ -109,83 +129,104 @@ impl TouchInternal {
...
+        unreachable!("There should be no wl_touch requests");
```
Comment (i509VCB) [inline src/wayland/seat/touch.rs:219]: Sorry I have made a mistake, there is a destructor. Let's just put a comment about that.

File: src/wayland/seat/mod.rs (hunk 2 above)
Comment (PolyMeilex) [reply src/wayland/seat/mod.rs]: Done

File: src/wayland/seat/mod.rs (hunk 5 above)
Comment (PolyMeilex) [reply src/wayland/seat/mod.rs]: Indeed

**Review by i509VCB — APPROVED** [review]

**Pull request merged**